File: scripts/validate_regression_publication_exports.py
"""Inspect the actual PDF and HWPX outputs of the publication-style fixtures."""
from pathlib import Path
from zipfile import ZipFile
import json
import logging
import re
from lxml import etree
import pdfplumber
logging.getLogger("pdfminer.pdffont").setLevel(logging.ERROR)
root = Path("tmp/regression-publication-style")
def normalize(text):
    return re.sub(r"\s+", "", text)
for mode in ("current", "accumulated"):
    expected = json.loads((root / f"{mode}-expected.json").read_text(encoding="utf-8"))
    with pdfplumber.open(root / f"{mode}.pdf") as pdf:
        assert "StatEdu" in (pdf.pages[0].extract_text() or ""), "PDF cover is missing"
        assert "Variable" not in (pdf.pages[0].extract_text() or ""), "Results overlap the cover"
        # PDF content order preserves wrapped cells; geometric line order interleaves adjacent columns.
        text = normalize("".join(page.extract_text(use_text_flow=True) or "" for page in pdf.pages))
        for value in expected:
            assert normalize(value) in text, (mode, value)
        assert any(page.width > page.height for page in pdf.pages)
    with ZipFile(root / f"{mode}.hwpx") as archive:
        tables = []
        for name in archive.namelist():
            if re.fullmatch(r"Contents/section\d+\.xml", name):
                tables += etree.fromstring(archive.read(name)).xpath("//*[local-name()='tbl']")
        widths = [float(n.get("width")) for n in tables[2].xpath("./*[local-name()='tr'][1]/*[local-name()='tc']/*[local-name()='cellSz']")]
        assert abs(widths[0]/sum(widths)-.28) < .005
        assert abs(widths[1]/sum(widths)-.16) < .005
    print(f"PASS: {mode} PDF headings/cells/notes, landscape page and HWPX conditional widths")

def word_signature(path):
    with ZipFile(path) as archive:
        doc = etree.fromstring(archive.read("word/document.xml"))
        ns = {"w": "http://schemas.openxmlformats.org/wordprocessingml/2006/main"}
        return (normalize("".join(doc.xpath("//w:t/text()",namespaces=ns))),
                [normalize("".join(t.xpath(".//w:t/text()",namespaces=ns))) for t in doc.xpath("//w:tbl",namespaces=ns)],
                [(n.get("{"+ns["w"]+"}w"),n.get("{"+ns["w"]+"}h")) for n in doc.xpath("//w:pgSz",namespaces=ns)])
assert word_signature(root / "accumulated.docx") == word_signature(root / "roundtrip.docx")
for name in ("current.docx", "accumulated.docx", "roundtrip.docx"):
    with ZipFile(root / name) as archive:
        doc = etree.fromstring(archive.read("word/document.xml"))
        body = list(doc.find("{*}body"))
        tables = [i for i,n in enumerate(body) if etree.QName(n).localname == "tbl"]
        between = body[tables[0]+1:tables[1]]
        paragraphs = [n for n in between if etree.QName(n).localname == "p"]
        text = lambda n: "".join(n.xpath(".//*[local-name()='t']/text()"))
        assert len(paragraphs) == 2 and "HC3 SE" in text(paragraphs[0]) and text(paragraphs[1]) == "", name
print("PASS: PDF cover and exactly one blank paragraph after table notes in Word and reopened HWPX")
print("PASS: Hancom reopen preserves complete text, table order and section dimensions")

File: scripts/validate_regression_publication_exports.R
source("scripts/validate_regression_publication_style.R", encoding="UTF-8")
entries <- c(entries, list(list(id="hierarchical", title="Hierarchical regression", html=hierarchical)))
normalize <- function(x) gsub("[[:space:]\u00a0]+", "", paste(x, collapse=""), perl=TRUE)
for (mode in c("current", "accumulated")) {
  selected <- if(mode=="current") list(list(id="current",title="Current results",html=paste(vapply(entries, `[[`, character(1), "html"),collapse="\n"))) else entries
  stem <- file.path(out,mode)
  write_result_collection_html(selected,paste0(stem,".html"))
  write_result_collection_docx(selected,paste0(stem,".docx"))
  save_result_collection_excel_file(selected,paste0(stem,".xlsx"))
  write_result_collection_pdf(selected,paste0(stem,".pdf"))
  write_result_collection_hwpx(selected,paste0(stem,".hwpx"))
  expected <- unlist(lapply(selected,function(entry) {
    doc <- xml2::read_html(entry$html)
    xml2::xml_text(xml2::xml_find_all(doc,"//th|//td|//h3|//h4|//*[contains(@class,'coefficient-note')]"))
  }))
  expected <- expected[nzchar(trimws(expected))]
  word <- xml2::read_xml(unz(paste0(stem,".docx"),"word/document.xml"))
  word_text <- normalize(xml2::xml_text(word))
  members <- unzip(paste0(stem,".hwpx"),list=TRUE)$Name
  hwpx_text <- normalize(vapply(members[grepl("Contents/section[0-9]+[.]xml$",members)],function(s) xml2::xml_text(xml2::read_xml(unz(paste0(stem,".hwpx"),s))), character(1)))
  wb <- openxlsx::loadWorkbook(paste0(stem,".xlsx"))
  excel_text <- normalize(unlist(lapply(seq_along(names(wb)),function(i) as.matrix(openxlsx::read.xlsx(wb,sheet=i,colNames=FALSE)))))
  for(value in expected) for(actual in list(word_text,hwpx_text,excel_text)) stopifnot(grepl(normalize(value),actual,fixed=TRUE))
  grids <- xml2::xml_find_all(word,".//w:tblGrid",xml2::xml_ns(word))
  stopifnot(length(grids)==length(names(wb)))
  conditional_widths <- as.numeric(xml2::xml_attr(xml2::xml_children(grids[[3]]),"w"))
  stopifnot(abs(conditional_widths[1]/sum(conditional_widths)-.28)<.005,abs(conditional_widths[2]/sum(conditional_widths)-.16)<.005)
  stopifnot(any(xml2::xml_attr(xml2::xml_find_all(word,".//w:pgSz",xml2::xml_ns(word)),"orient")=="landscape",na.rm=TRUE))
  jsonlite::write_json(expected,paste0(stem,"-expected.json"),auto_unbox=FALSE)
  message("PASS: ",mode," five exports; all table cells, headings/subtitles, notes, conditional widths and landscape section")
}
